File: spec/javascript/app/views/login_spec.js
//= require "../../hyperarchy_spec_helper"

Screw.Unit(function(c) { with(c) {
  describe("Views.Login", function() {
    var view;
    before(function() {
      view = Views.Login.to_view();
    });


    describe("when #login_submit is clicked", function() {
      before(function() {
        view.find('#email_address').val("cobham@gmail.com");
        view.find('#password').val("spectrum");

        expect(Application.posts).to(be_empty);
        view.find('#login_submit').click();

      });

      it("posts the form field values to /login", function() {
        expect(Application.posts.length).to(equal, 1);
        expect(Application.last_post.url).to(equal, "/login");
        expect(Application.last_post.data).to(equal, {
          email_address: "cobham@gmail.com",
          password: "spectrum"
        });
      });


      context("when the result is successful", function() {
        it("sets the Application.current_user_id and navigates to #elections if the result is successful", function() {
          mock(jQuery.history, 'load', function() {
            expect(Application.current_user_id).to(equal, "billy");
          });
          Application.last_post.simulate_success({
            current_user_id: "billy"
          });
          expect(jQuery.history.load).to(have_been_called, with_args("elections"));
        });
      })

      context("when the result is unsuccessful", function() {
        it("displays the error and does not set the Application.current_user_id or navigate to #elections", function() {
          mock(jQuery.history, 'load');
          Application.last_post.simulate_failure({
            errors: {
              password: "Your password did not match the given email address."
            }
          });
          expect(jQuery.history.load).to_not(have_been_called);
          expect(view.html()).to(match, "Your password did not match the given email address.");
        });
      })

    });
  });
}});

File: server/spec/hyperarchy/rack/login_spec.rb
require File.expand_path("#{File.dirname(__FILE__)}/../../hyperarchy_spec_helper")

describe "POST /login", :type => :rack do
  attr_reader :user

  before do
    @user = User.create(:email_address => "billy@example.com", :password => "spectrum")
  end

  context "if a User with the given email address exists" do
    context "if the given password matches that User" do
      it "sets the current_user and returns a successful ajax response with the current_user_id and the user record" do
        post "/login", :email_address => user.email_address, :password => "spectrum"

        last_response.should be_ok
        last_response.body_from_json.should == {
          "successful" => true,
          "data" => {
            "current_user_id" => user.id
          },
          "dataset" => {
            "users" => {
              user.id.to_s => user.wire_representation
            }
          }
        }
      end
    end

    context "if the given password does NOT match that User" do
      it "does not set the current_user and returns an unsuccessful ajax response with errors on password" do
        post "/login", :email_address => user.email_address, :password => "incorrectpassword"

        last_response.should be_ok
        last_response.body_from_json.should == {
          "successful" => false,
          "data" => {
            "errors" => {
              "email_address" => nil,
              "password" => ["Incorrect password."]
            }
          }
        }
      end
    end
  end

  context "if no User with the given email address exists" do
    it "does not set the current_user and returns an unsuccessful ajax response with errors on email_address" do
      post "/login", :email_address => "bogus@example.com", :password => "spectrum"

      last_response.should be_ok
      last_response.body_from_json.should == {
        "successful" => false,
        "data" => {
          "errors" => {
            "email_address" => ["No user found with that email address."],
            "password" => nil
          }
        }
      }
    end
  end
end